src/engine/controls/keycontrol.cpp

```diff
@@ -256,6 +256,13 @@ void KeyControl::updateRate() {
     qDebug() << "      =" << qSetRealNumberPrecision(18) << m_pitchRateInfo.pitchRatio;
     qDebug() << "      =" << m_pitchRateInfo.pitchRatio;
 #endif
+    // If we just unlocked the original key with speed != +-1.0 we may encounter wrong
+    // decimals (e.g. 1 - 1.73123e-09) after pitchRatio calculation round-trip.
+    // Round to 1.0 to avoid false positive pitch offset (pitchOctaves != 0).
+    double pitchRatioDiffTo1 = fabs(1.0 - m_pitchRateInfo.pitchRatio);
+    if (0.0 < pitchRatioDiffTo1 && pitchRatioDiffTo1 < pow(10, -9)) { // 0.000000001
+        m_pitchRateInfo.pitchRatio = 1.0;
+    }
```

Comment:
I'm struggling to understand the reasoning behind this because I struggle to understand the surrounding code. `m_pitchRateInfo.pitchRatio` is basically imprecise because of rounding errors? But only around 1.0?

Reply:
Basically roundtrip rounding erros, yes.
Tbh I didn't manage to understand how all the ratios relate in all situations either because I'm still confused by the naming, but with debug output enabled it's at least easy to trace how the `m_pitchRateInfo` values change. In https://github.com/mixxxdj/mixxx/pull/4710#issuecomment-1088142652 I outlined the circumstances and the procedure to reproduce the imprecision.

Reply:
The issue is that the new statement in line 216 is designed to make m_pitchRateInfo.pitchRatio = 1, but it is done indirect via other variables. It would be straight forward and more easy to understand to set it 1 explicit. This would require more changes in the code though.